napari/_qt/qt_console.py

```diff
@@ -98,7 +94,5 @@ def __init__(self, user_variables=None):
         # self.execute_on_complete_input = True
 
     def shutdown(self):
-        if self.kernel_client is not None:
-            self.kernel_client.stop_channels()
-        if self.kernel_manager is not None:
-            self.kernel_manager.shutdown_kernel()
+        self.kernel_client.stop_channels()
+        self.kernel_manager.shutdown_kernel()
```

Comment:
Are you sure this change is correct? both of these can be None if the viewer is run in an IPython shell (see line 61):

```python
        elif isinstance(shell, TerminalInteractiveShell):
            # if launching from an ipython terminal then adding a console is
            # not supported. Instead users should use the ipython terminal for
            # the same functionality.
            self.kernel_client = None
            self.kernel_manager = None
            self.shell = None
            self.push = lambda var: None
```

Having said that, I just tested this, and I didn't manage to make it crash (I couldn't access the QtConsole object to use this method)...

Reply:
Yea it won't crash but ideally we should also be shutting down the kernal with both the TerminalInteractiveShell and the ZMQInteractiveShell, so that's my next investigation. The updates here will fix the issues with the ipykernel but I don't think it's a full solution yet.

Reply:
No, the point with the TerminalInteractiveShell is that there is no kernel! And with ZMQ there is one but I don't think we should shut it down: napari might be one of many clients connecting to that kernel.

Reply:
oh great then!

Reply:
I was just about to open a PR about this and then found this conversation.  I'm not sure i follow what the rationale was behind removing the `if self.kernel_client is not None` conditional.
I'm getting an error now in the master branch whenever I close a viewer (when starting in ipython):

```python
In [1]: import napari, numpy as np

In [2]: v = napari.view_image(np.random.rand(10,10,10,10))

# upon closing the window

Traceback (most recent call last):
  File "/Users/talley/Dropbox (HMS)/Python/forks/napari/napari/_qt/qt_main_window.py", line 324, in closeEvent
    self.qt_viewer.console.shutdown()
  File "/Users/talley/Dropbox (HMS)/Python/forks/napari/napari/_qt/qt_console.py", line 97, in shutdown
    self.kernel_client.stop_channels()
AttributeError: 'NoneType' object has no attribute 'stop_channels'
```